### DISCUSSION TITLE
LiteLLM Example Configs

### DISCUSSION THREAD

**Discussion by justinh-rahb** [open-webui/open-webui#1038] [Show and tell]
> [!CAUTION]  
> # NONE OF THE BELOW IS RELEVANT ANYMORE, LITELLM HAS BEEN REMOVED FROM OPEN WEBUI

https://docs.openwebui.com/migration#migrating-from-internal-to-external-litellm

<details>
Just wanted to dump some configs for common endpoints for future reference.

> [!NOTE]
> Other than official OpenAI-endpoints, LiteLLM usually requires that an `provider` be specified before the `model string` in the `Add a model` field, example:
> ```
> azure/gpt35turbo
> ```
> `Model Name` can be whatever you want it to appear as in your list.
> `API Base URL` typically only needs to be set for OpenAI-compatible APIs and Azure.

> [!WARNING]
> #### Gemini Endpoint:
> **ONLY** Makersuite/AI Studio API keys are supported, VertexAI/GCP endpoints and authentication methods are **NOT** currently supported.

## Anthropic
Model strings:
- `claude-2`
- `claude-2.1`
- `claude-instant-1.2`
- `claude-3-sonnet-20240229`
- `claude-3-opus-20240229`
- `claude-3-haiku-20240307`

> [!NOTE]
> Anthropic's API requires that `max_tokens` parameter be sent in the payload, the maximum accepted value is `4096`

### Claude 2.1
![Screenshot 2024-03-04 at 3 42 17 PM](https://github.com/open-webui/open-webui/assets/52832301/39cfeb61-3186-44ba-9d07-756a20187bd5)
### Claude 3 "Sonnet"
![Screenshot 2024-03-04 at 3 33 12 PM](https://github.com/open-webui/open-webui/assets/52832301/8dbc671f-80cc-403b-bb1d-999d7ecb4002)
### Claude 3 "Opus"
![Screenshot 2024-03-04 at 3 33 02 PM](https://github.com/open-webui/open-webui/assets/52832301/343223f5-c1fc-4c6c-a3fe-e93c12531d40)

## Groq
Model strings:
- `mixtral-8x7b-32768`
- `llama2-70b-4096`
### Mixtral 8x7B
![Screenshot 2024-03-04 at 3 33 39 PM](https://github.com/open-webui/open-webui/assets/52832301/1e9de3b3-7574-4e3b-bd87-a54406af94b9)
### Llama2 70B
![Screenshot 2024-03-04 at 3 33 53 PM](https://github.com/open-webui/open-webui/assets/52832301/8db1bd01-d52c-4e19-8139-cec7fe4e1f8e)

## Google Gemini
Model strings:
- `gemini-pro`
### Gemini Pro
![Screenshot 2024-03-04 at 3 34 24 PM](https://github.com/open-webui/open-webui/assets/52832301/8267c8b0-2c75-4ec9-a5df-4b226e1728af)

## Mistral
Model strings:
- `open-mistral-7b` (aka `mistral-tiny-2312`)
- `open-mixtral-8x7b` (aka `mistral-small-2312`)
- `mistral-small-latest` (aka `mistral-small-2402`)
- `mistral-medium-latest` (aka `mistral-medium-2312`)
- `mistral-large-latest` (aka `mistral-large-2402`)
### Open Mixtral 8x7B (formerly `mistral-small`)
![Screenshot 2024-03-04 at 3 48 24 PM](https://github.com/open-webui/open-webui/assets/52832301/0d27bb5b-dc6f-4006-bf7b-70027e59cde7)
### Mistral Medium
![Screenshot 2024-03-04 at 3 46 27 PM](https://github.com/open-webui/open-webui/assets/52832301/50b59f6b-bead-4e20-9533-88b03b9bda9f)
### Mistral Large
![Screenshot 2024-03-04 at 3 38 45 PM](https://github.com/open-webui/open-webui/assets/52832301/f1d4754a-bed4-42bf-a4ae-347e041eb275)

## Azure OpenAI
Model strings:
- `gpt35turbo`
- `gpt4`
### GPT 3.5 Turbo
![Screenshot 2024-03-04 at 3 36 01 PM](https://github.com/open-webui/open-webui/assets/52832301/881cd44d-735b-44ce-8d68-a255b3a32aeb)

---

> [!CAUTION]
> ## OpenAI and Ollama Endpoints
> Unless there is a specific reason why the **Connections > Ollama** or **Connections > OpenAI** methods are not working for your use-case, the following methods are **NOT** recommended to be used instead:

## OpenAI
Model strings:
- `gpt-3.5-turbo`
- `gpt-4`
- `gpt-4-turbo-preview`
- `gpt-4-vision-preview`
- [See more](https://platform.openai.com/docs/models)
### GPT 4 Turbo
<img width="694" alt="Screenshot 2024-03-06 at 12 03 03 AM" src="https://github.com/open-webui/open-webui/assets/52832301/7afafb20-bf01-42a8-8fbf-f381a54425c9">

### "OpenAI-compatible" endpoints
![Screenshot 2024-03-04 at 3 38 07 PM](https://github.com/open-webui/open-webui/assets/52832301/29c10559-ce2c-4860-b02f-7b09cc7087ae)

### Llama2
![Screenshot 2024-03-04 at 3 37 00 PM](https://github.com/open-webui/open-webui/assets/52832301/b2fae8d8-d12f-4d70-b53c-d5bd2dcade41)
</details>

**Comment by HydeAndGeek** [comment]
ahhhh yes ty!

**Comment by PierrunoYT** [comment]
Hey it's not working 
![Screenshot 2024-03-10 101727](https://github.com/open-webui/open-webui/assets/95778421/b04f7adb-fe9f-4c25-923c-d1b31e06a8d4)

**Reply by PierrunoYT** [reply]
Sorry I'm just confused because Pinokio says it's up to date.

**Comment by ghost** [comment]
this is amazing @justinh-rahb - if you have any feedback for how we can improve our own docs for this, let me know - https://docs.litellm.ai/docs/

- Krrish, LiteLLM Maintainer

**Reply by ghost** [reply]
btw if you're on linkedin/discord, happy to set up a direct support channel as well

- LinkedIn: https://www.linkedin.com/in/krish-d/
- Discord: krrish8748

**Reply by justinh-rahb** [reply]
@krrishdholakia nice to hear from you guys, and thanks for this fantastically useful swiss-army knife library! Personally, I found the documenation around using the library with Python apps to be quite thorough, but I've had to read between the lines and compose my solutions for the proxy server configuration by my own interpretation at times. I was planning on re-writing these docs above to be more readable and put them on our own docs site, feel free to share our notes if you like!

**Reply by abdimussa87** [reply]
I get an error when I try to use claude-3-opus-20240229 & claude-3-sonnet-20240229:
```
AnthropicException - AnthropicException - {
"error": {
"type": "forbidden",
"message": "Request not allowed"
}
}
```

**Reply by justinh-rahb** [reply]
> I get an error when I try to use claude-3-opus-20240229 & claude-3-sonnet-20240229:
> 
> ```
> AnthropicException - AnthropicException - {
> "error": {
> "type": "forbidden",
> "message": "Request not allowed"
> }
> }
> ```

My guess is invalid API key.

**Reply by ghost** [reply]
@abdimussa87 if you run the proxy with `--debug` you should be able to see the request being sent to anthropic.

**Reply by abdimussa87** [reply]
It had to do with the region I was in. Now I'm getting this error:
`AnthropicException - AnthropicException - {"type":"error","error":{"type":"invalid_request_error","message":"\"claude-3-opus-20240229\" is not supported on this API. Please use the Messages API instead."}}`

**Reply by justinh-rahb** [reply]
> It had to do with the region I was in. Now I'm getting this error: `AnthropicException - AnthropicException - {"type":"error","error":{"type":"invalid_request_error","message":"\"claude-3-opus-20240229\" is not supported on this API. Please use the Messages API instead."}}`

You're running an outdated version of Open WebUI container with an older version of LiteLLM prior to the release of Claude 3 API.

**Reply by abdimussa87** [reply]
Yup. That was it. Thank you.

**Comment by yelban** [comment]
No, in v0.1.115 (latest version) open-webui still cannot set LiteLLM to make any Claude 3 model work. Unless...

# Upgrade LiteLLM to the latest version v1.34.12.

`git clone https://github.com/open-webui/open-webui.git`

`cd open-webui`

Replace litellm==1.30.7 with litellm==1.34.12 in the ./backend/requirements.txt file.

`docker build -t ghcr.io/open-webui/open-webui:latest .`

Then use the locally created docker image ghcr.io/open-webui/open-webui:latest.

You can use the Claude 3 models added by LiteLLM.

**Reply by justinh-rahb** [reply]
Look at this, all workin' and stuff:

<img width="684" alt="Screenshot 2024-03-29 at 10 55 05 AM" src="https://github.com/open-webui/open-webui/assets/52832301/a8727371-17c7-4b63-afb2-7a211d750a98">

**Reply by yelban** [reply]
May I ask if you are also using the Docker container to start open-webui?

When I used the official mirror: ghcr.io/open-webui/open-webui:latest, following the previous example, carefully checking and adding claude 3 litellm has never been successful until modifying the version of litellm in requirements.txt and using a self-built docker image to finally succeed.

**Reply by justinh-rahb** [reply]
> May I ask if you are also using the Docker method to start open-webui?

Docker-installed `main` branch - v0.1.115, no modifications.

**Comment by yelban** [comment]
Delete the local open-webui images, After restarting using ghcr.io/open-webui/open-webui:main, I can confirm that the newly added claude 3 model can run correctly.

`Docker-installed main branch - v0.1.115, no modifications` It works perfectly.

Your response and reminder are greatly appreciated. Thank you. @justinh-rahb

**Comment by flefevre** [comment]
Thanks a lot for the elements you shared. I am facing a difficulty in setting the access of open webui to litellm.
I do not understand what i need to provide in the litellm-config.yaml file for the sk.
Should i connect to litellm to generate a specific key and put it in master_key?
Thanks for the explanoation.


`openwebui  |   File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 372, in user_api_key_auth
openwebui  |     assert api_key.startswith("sk-")  # prevent token hashes from being used
openwebui  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
openwebui  | AssertionError
openwebui  | id='dccf746e-2efa-4248-923a-7bfddedea4e4' name='Iagen Innov' email='tot@email.com' role='admin' profile_image_url='/user.png' timestamp=1711742369
openwebui  | INFO:     172.22.0.1:0 - "GET /litellm/api/model/info HTTP/1.1" 401 Unauthorized
`

**Reply by justinh-rahb** [reply]
Master key is irrelevant in the context of our internal LiteLLM proxy! You seem to be wanting to setup a separate external LiteLLM as it's own container. Use it as if it is an OpenAI endpoint, where Master key is your `OPENAI_API_KEY` you would use for the "OpenAI Connection", with base URL set to your LiteLLM container's address.

**Reply by flefevre** [reply]
Thanks for your answer.
You are right. I have installed in two separate docker compose files: 1) open webui and 2) litellm with redis.
We have had an exchange about it here with you and shared it here to help people.: [https://github.com/open-webui/open-webui/discussions/508#discussioncomment-9051366](https://github.com/open-webui/open-webui/discussions/508#discussioncomment-9051366)

**Comment by zhixwang** [comment]
Hi, thanks for the support! But I'm not clear how to set it up on OpenWebUI. It there a step-by-step tutorial somewhere? :)

**Comment by ravidreams** [comment]
How to get the LiteLLM API key? What is the LiteLLM API base URL? Are we supposed to install LiteLLM before adding values here? How to install that and get it connected with Openwebui? WHat value to give for RPM?

**Comment by flefevre** [comment]
Dear all,
If I do understand,open webui comes with its own litellm. It is built
inside.
Perhaps they could allow at term to reference an external litellm, a better
practice.

François

**Comment by JuanjoJG-WATA** [comment]
I recently updated OUI and I tried to add my Azure API Key and endpoint but with no results. The LiteLLM option there used to be is nowhere to be seen

**Reply by justinh-rahb** [reply]
Here's a guide to help you move your `litellm` setup out of the `open-webui` container and into its own separate container. This will allow you to use `litellm` independently and connect it to `open-webui` as needed.

1. **Download your `config.yaml` file**

Your `config.yaml` file can be downloaded from the Admin Settings window:

![Screenshot 2024-06-04 at 10 34 27 AM](https://github.com/open-webui/open-webui/assets/52832301/ecb71d92-2aeb-444d-a00f-024db4789333)

2. **Run a standalone `litellm` container**

To run `litellm` in its own container, use the following `docker run` command:
```bash
docker run -d \
  -p 4000:4000 \
  --name litellm \
  -v ./config.yaml:/app/config.yaml \
  -e LITELLM_MASTER_KEY=sk-12345 \
  --restart always \
  ghcr.io/berriai/litellm:main-latest \
  --config /app/config.yaml --port 4000
```
Replace `./config.yaml` with the path to the `config.yaml` file you downloaded. Set the `LITELLM_MASTER_KEY` environment variable to any API key value as long as it's not empty.

3. **Connect `litellm` to `open-webui`**

Once you have the `litellm` container running, you can connect it to the `open-webui` container. In the `open-webui` settings, add an "OpenAI" base URL under "Connections" with the value `http://host.docker.internal:4000/v1` and set the API key to your `LITELLM_MASTER_KEY` value (e.g., `sk-12345`).

By following these steps, you'll have moved your `litellm` setup out of the `open-webui` container and into its own separate container. You can now use `litellm` independently and connect it to `open-webui` as needed.

**Reply by JuanjoJG-WATA** [reply]
Thank you for your answer Justin. Unfortunately, I'm not using Docker, is there a way to do so without it?

**Reply by justinh-rahb** [reply]
Unfortunately, no, not a straightforward one. You could consider our own alternative to litellm, [Pipelines](https://github.com/open-webui/pipelines) which admittedly is not at parity with litellm but we can iterate on it very rapidly. There is an Azure API example though it's possible you may need to help us debug it since afaik none of us devs have access to test it ourselves.

**Reply by ghost** [reply]
hey @JuanjoJG-WATA yes you can use litellm without a docker container - via the cli tool 

1. Install litellm 
```
pip install 'litellm[proxy]' 
```

2. Run with config 
```
litellm --config /path/to/config.yaml

# RUNNING on http://0.0.0.0:4000
```

docs - https://docs.litellm.ai/docs/proxy/quick_start

**Reply by JuanjoJG-WATA** [reply]
I made it work after a few tweaks. @justinh-rahb @krrishdholakia @tjbck 

First I installed litellm like Krish said in the comment above this one:
`pip install 'litellm[proxy]' `

Then I changed directory to open-webui and created a yaml file called **config.yaml** and I filled it with the first 6 lines of this example:
(https://docs.litellm.ai/docs/proxy/quick_start#create-a-config-for-litellm-proxy)

Then I ran the .yaml with:
`litellm --config open-webui/config.yaml`

In the OpenWebUI interface I went to Settings -> Connections -> Made a new OpenAI API with **http://localhost:4000** and also tried with the proper IP, alfo followed by the port to make sure it worked remotely.

Also I made some changes to the .env file. To that I added the following:
![image](https://github.com/open-webui/open-webui/assets/165764759/c02546cf-c848-4d70-beee-7fce9971c98e)

I closed both the process of the yaml and openwebui and ran them again to commit the changes.

And voilá, it's working.

![image](https://github.com/open-webui/open-webui/assets/165764759/26f6b4dc-99d4-46b3-8dd8-b4c7ece1ff17)

**Comment by flefevre** [comment]
Please have a look to the new version>0.2

Use environmental variable
https://docs.openwebui.com/getting-started/env-configuration#litellm

Le jeu. 6 juin 2024, 09:24, JuanjoJG-WATA ***@***.***> a
écrit :

**Reply by JuanjoJG-WATA** [reply]
Yes, I forgot to include that in my answer.

**Comment by Fusseldieb** [comment]
I just realized that this is a minor step back. It shouldn't have been removed just yet.

I've had Anthropic and OpenAI added as LiteLLM models, and it worked great, for both image and text, however that's not working anymore.

I now have three options:
- Move to Pipelines, which don't support images yet, therefore loosing image capabilities of _BOTH_ models
- Stay on the vanilla OpenAI connection, but loosing Anthropic
- Downgrade

Honestly, all three don't sound great haha

**Reply by justinh-rahb** [reply]
You've forgotten a 4th option:

* Keep using LiteLLM if that's working for you

The only thing we've "taken away" is the UI for configuring it, but good news! LiteLLM has their own.

**Reply by Fusseldieb** [reply]
Hmm... Didn't knew that. 
I run open-webui inside a docker container, so I'm unsure where I would configure it.
Would you mind explaining? Might even benefit some other readers which have the same "issue", until the gap closes.

**Reply by justinh-rahb** [reply]
https://docs.openwebui.com/migration#migrating-from-internal-to-external-litellm

https://github.com/open-webui/open-webui/discussions/1038#discussioncomment-9665854
Was also posted 3 days ago, in this thread you are currently reading ☝️

**Comment by chuckjewell** [comment]
Previously had claude-3 working with the integrated litellm and followed the directions in this thread for running it on docker but having issues connecting open-webui to the litellm server.

- I've updated to the newest version of open-webui using the docker update command in the docs
- Installed the litellm docker container and verified it is working and can list my models and do completions using my claude-3 config
- Have an https://api.openai.com/v1 in the same dialog also configured that works great.

Here is me hitting litellm with my claude-3 config and getting a completion:
```
curl -X 'POST' \
  'http://localhost:4000/completions?model=claude-3' \
  -H 'accept: application/json' \
  -H 'API-Key: My-LiteLLM-API-KEY' \
  -H 'Content-Type: application/json' \
  -d '{
    "prompt": "Once upon a time",
    "max_tokens": 50
  }'
{"id":"chatcmpl-b891a8d1-f6a0-4381-8962-35f6e09971c9","object":"text_completion","created":1718731082,"model":"claude-3-opus-20240229","choices":[{"finish_reason":"length","index":0,"text":"There was a curious young girl named Lily who loved to explore the enchanted forest near her village. Every day, she would venture into the woods, marveling at the towering trees, the vibrant flowers, and the gentle whispers","logprobs":null}],"usage":{"prompt_tokens":11,"completion_tokens":50,"total_tokens":61}}
```

The issue I am seeing is when I add it as an OpenAI API in the open-webui settings (using http://host.docker.internal:4000/v1 and my litellm secret) I get an "OpenAI Network Problem" error in the UI when I hit the 'verify connection button'.

I have tried changing the address to http://localhost:4000/v1, removing the /v1 (doesn't seem to show v1 in the litellm swagger).

Just not sure why they can't talk. Any ideas?


Edit: Got it working by creating a docker network and adding the containers to that network and using the litellm IP address. I'm sure there is a more elegant solution but here is what worked for me:
```
docker rm -f litellm open-webui watchtower_webui

docker network create litellm_network


docker run -d \
  --network litellm_network \
  --name litellm \
  -p 4000:4000 \
  -v ./config.yaml:/app/config.yaml \
  -e LITELLM_MASTER_KEY=ba5ed4e8-05eb-4c3b-92f9-fbec67f35a57 \
  --restart always \
  ghcr.io/berriai/litellm:main-latest \
  --config /app/config.yaml --port 4000


  docker run -d \
  --network litellm_network \
  -p 3000:8080 \
  -v ollama:/root/.ollama \
  -v open-webui:/app/backend/data \
  --name open-webui \
  --restart always ghcr.io/open-webui/open-webui:ollama
  
  ```

**Reply by OM-EL** [reply]
you need to change this "http://localhost:4000/v1" to "http://host.docker.internal:4000/v1 " @chuckjewell
